### PR TITLE
VReplication: Improve Error/Status Reporting

### PR DESCRIPTION
**Pull request by mattlord**
## Description

This PR addresses the issues raised in [the bug report](https://github.com/vitessio/vitess/issues/12051) along with some other related improvements that I saw were needed as I worked on [the new troubleshooting guide](https://deploy-preview-1333--vitess.netlify.app/docs/16.0/user-guides/migration/troubleshooting/):
1. Adds unit tests for the `Progress` (which is a parent of `Show`) action
2. When the stream is in an error state, show the error message in the `Show`/`Progress` action output. For example:
    ```
    $ vtctlclient MoveTables -- Progress customer.commerce2customer
    The following vreplication streams exist for workflow customer.commerce2customer:

    id=1 on 0/zone1-0000000201: Status: Error: Duplicate entry '6' for key 'customer.PRIMARY' (errno 1062) (sqlstate 23000) during query: insert into customer(customer_id,email) values (6,'mlord@planetscale.com').
    ```
3. Adds additional errors to the [unRecoverableError() error list](https://github.com/vitessio/vitess/pull/12052/files#diff-bbb27b541b1028d8ef34e417eb9b4fbd974e7903e927670963a345897b206320) as things like `"Unknown column 'notes' in 'field list' (errno 1054) (sqlstate 42S22) during query: insert into customer(customer_id,email,notes) values (100,'test@tester.com','Lots of notes')"` were not covered AND keeping in mind that [users can now modify the `on-ddl` value](https://github.com/vitessio/vitess/pull/11532) and DDL may come through the stream

<details>
<summary>Click here to see new and improved outputs from the test case in the bug report:</summary>

```
$ vtctlclient MoveTables -- --tablet_types=RDONLY --source commerce --tables 'customer,corder' Create customer.commerce2customer
Waiting for workflow to start:
0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ... 0% ...
The workflow did not start within 30s. The workflow may simply be slow to start or there may be an issue.
Check the status using the 'Workflow customer.commerce2customer show' client command for details.
MoveTables Error: rpc error: code = Unknown desc = timed out waiting for workflow to start
E0106 15:09:23.724578   99838 main.go:105] remote error: rpc error: code = Unknown desc = timed out waiting for workflow to start
```

```
$ vtctlclient Workflow -- customer.commerce2customer show
{
	"Workflow": "commerce2customer",
	"SourceLocation": {
		"Keyspace": "commerce",
		"Shards": [
			"0"
		]
	},
	"TargetLocation": {
		"Keyspace": "customer",
		"Shards": [
			"0"
		]
	},
	"MaxVReplicationLag": 51,
	"MaxVReplicationTransactionLag": 1673036368,
	"Frozen": false,
	"ShardStatuses": {
		"0/zone1-0000000200": {
			"PrimaryReplicationStatuses": [
				{
					"Shard": "0",
					"Tablet": "zone1-0000000200",
					"ID": 1,
					"Bls": {
						"keyspace": "commerce",
						"shard": "0",
						"filter": {
							"rules": [
								{
									"match": "customer",
									"filter": "select * from customer"
								},
								{
									"match": "corder",
									"filter": "select * from corder"
								}
							]
						}
					},
					"Pos": "",
					"StopPos": "",
					"State": "Error",
					"DBName": "vt_customer",
					"TransactionTimestamp": 0,
					"TimeUpdated": 1673036317,
					"TimeHeartbeat": 0,
					"TimeThrottled": 0,
					"ComponentThrottled": "",
					"Message": "Error picking tablet: context has expired",
					"Tags": "",
					"WorkflowType": "MoveTables",
					"WorkflowSubType": "None",
					"CopyState": null
				}
			],
			"TabletControls": null,
			"PrimaryIsServing": true
		}
	},
	"SourceTimeZone": "",
	"TargetTimeZone": ""
}
```

```
# When we're trying to start, but have not yet produced an error or made any progress
$ vtctlclient MoveTables -- Progress customer.commerce2customer
The following vreplication streams exist for workflow customer.commerce2customer:

id=1 on 0/zone1-0000000200: Status: Running. VStream has not started.


# When we've encountered an error
$ vtctlclient MoveTables -- Progress customer.commerce2customer
The following vreplication streams exist for workflow customer.commerce2customer:

id=1 on 0/zone1-0000000201: Status: Error: Error picking tablet: context has expired.


# When the copy phase is actually in progress
$ vtctlclient MoveTables -- Progress customer.commerce2customer
Copy Progress (approx):

corder: rows copied 0/5 (0%), size copied 16384/16384 (100%)
customer: rows copied 0/5 (0%), size copied 16384/16384 (100%)


The following vreplication streams exist for workflow customer.commerce2customer:

id=1 on 0/zone1-0000000200: Status: Copying. VStream Lag: 0s.


# When the copy phase is actually completed
$ vtctlclient MoveTables -- Progress customer.commerce2customer
The following vreplication streams exist for workflow customer.commerce2customer:

id=1 on 0/zone1-0000000200: Status: Running. VStream Lag: 0s.
```
</details>

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/12051

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added
-   [x] Documentation was added in [the User Guides update](https://github.com/vitessio/website/pull/1333/files/aa36bbef9d7d2d74dd790603b994e8260f908f41#diff-110b19887dc4ec2f46980f48015968c11a1b62cd26bd0a0746a3e3adc170f1f4R280-R292)

### CODE REVIEW

File: go/mysql/constants.go [hunk 1]
```diff
@@ -368,6 +368,7 @@ const (
 	ERServerShutdown = 1053
 
 	// not found
+	ERDbDropExists          = 1008
```
Comment (mattlord) [inline go/mysql/constants.go:371]: Changes in this file were just me ordering the subgroups by error code as it makes it easier to walk through all of them as I did to see what ones should be in the `vreplication.unRecoverable()` list.

File: go/vt/vttablet/tabletmanager/vreplication/controller.go [hunk 2]
```diff
@@ -206,7 +206,11 @@ func (ct *controller) runBlp(ctx context.Context) (err error) {
 	var tablet *topodatapb.Tablet
 	if ct.source.GetExternalMysql() == "" {
 		log.Infof("trying to find a tablet eligible for vreplication. stream id: %v", ct.id)
-		tablet, err = ct.tabletPicker.PickForStreaming(ctx)
+		// Retry 5 times before giving up and returning an error that the user
+		// can see and act upon if needed.
+		tpCtx, tpCancel := context.WithTimeout(ctx, discovery.GetTabletPickerRetryDelay()*5)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/vreplication/controller.go:211]: let's make that `5` a `const` variable  at the top of this file

File: go/vt/vtctl/vtctl.go [hunk 3]
```diff
@@ -2419,9 +2421,13 @@ func commandVRWorkflow(ctx context.Context, wr *wrangler.Wrangler, subFlags *pfl
 					return nil
 				}
 				wr.Logger().Printf("%d%% ... ", 100*progress.started/progress.total)
+			case <-timedCtx.Done():
+				wr.Logger().Printf("\nThe workflow did not start within %s. The workflow may simply be slow to start or there may be an issue.\n",
+					(*timeout).String())
+				wr.Logger().Printf("Check the status using the 'Workflow %s show' client command for details.\n", ksWorkflow)
```
Comment (shlomi-noach) [inline go/vt/vtctl/vtctl.go:2427]: I wonder if it possible to check the status at this point and print the output? Save the user an extra step?

**Review by shlomi-noach — APPROVED** [review]
LGTM, a couple minor comments at your discretion

**Comment by shlomi-noach** [comment]
While ticking the review template, I noted

>  If a test is added or modified, there should be a documentation on top of the test to explain what the expected behavior is what the test does.

Could you please do that for the new `TestMoveTables`? (noting to myself to similarly document my tests)

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

File: go/vt/vttablet/tabletmanager/vreplication/controller.go (hunk 2 above)
Comment (mattlord) [reply go/vt/vttablet/tabletmanager/vreplication/controller.go:211]: Good idea. Done!

File: go/vt/vtctl/vtctl.go (hunk 3 above)
Comment (mattlord) [reply go/vt/vtctl/vtctl.go:2427]: IMO the output is already a little noisy and at this point (at least sub-second before) it had not yet started. I also like that we're informing/reminding them of the way to monitor the status and see the workflow details.

**Pull request merged**